File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.js
const sumFunc = (arr) => arr.reduce((acc, cur) => acc + cur, 0);

function solution(queue1, queue2) {
    let [answer, sum1, sum2, idx1, idx2] = [0, sumFunc(queue1), sumFunc(queue2), 0, 0];
    const [half, limit] = [(sum1 + sum2) / 2, queue1.length * 3];

    while (answer < limit) {
        if (sum1 === sum2)
            break;

        if (sum1 > half) {
            const data = queue1[idx1++];
            sum1 -= data, sum2 += data, queue2.push(data);
        } else {
            const data = queue2[idx2++];
            sum1 += data, sum2 -= data, queue1.push(data);
        }
        ++answer;
    }

    return (answer < limit) ? answer : -1;
}

File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>
#include <cmath>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = 0;
    long sum1 = ::accumulate(queue1.begin(), queue1.end(), 0),
           sum2 = ::accumulate(queue2.begin(), queue2.end(), 0);
    size_t idx1 = 0, idx2 = 0;
    int max = queue1.size() * 4;

    while (answer < max)
    {
        if (sum1 > sum2 && !queue1.empty())
        {
            queue2.emplace_back(queue1[idx1]);
            sum2 += queue1[idx1], sum1 -= queue1[idx1];
            ++idx1;
        }
        else if (sum1 < sum2 && !queue2.empty())
        {
            queue1.emplace_back(queue2[idx2]);
            sum1 += queue2[idx2], sum2 -= queue2[idx2];
            ++idx2;
        }
        else break;
        
        ++answer;
    }

    return answer >= max ? -1 : answer;
}